gtsfm/frontend/correspondence_augmenter.py

```diff
@@ -0,0 +1,179 @@
+from gtsam import Pose3, Rot3, Unit3
+
+import gtsfm.common.types as gtsfm_types
+import gtsfm.frontend.two_view_ba_utils as two_view_ba_utils
+from gtsfm.bundle.bundle_adjustment import BundleAdjustmentOptimizer
+from gtsfm.bundle.two_view_ba import TwoViewBundleAdjustment
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.common.pose_prior import PosePrior, PosePriorType
+from gtsfm.data_association.point3d_initializer import TriangulationOptions
+from gtsfm.frontend.correspondence_generator.correspondence_generator_base import CorrespondenceGeneratorBase
+from gtsfm.two_view_estimator import TWO_VIEW_OUTPUT, TWO_VIEW_REPORT_TAG, generate_two_view_report_from_result
+
+RELATIVE_POSE_PRIOR_SIGMA = np.ones((6,)) * 3e-2
+
+
+class CorrespondenceAugmenter:
+    """Augment correspondences on two-view estimates."""
+
+    def __init__(
+        self,
+        correspondence_generator: CorrespondenceGeneratorBase,
+        eval_threshold_px: float,
+        triangulation_options: TriangulationOptions,
+        bundle_adjust_2view_maxiters: int = 10,
+        ba_reproj_error_thresholds: List[Optional[float]] = [3],
+    ) -> None:
+        self._correspondence_generator = correspondence_generator
+        self._corr_metric_dist_threshold = eval_threshold_px
+        self._triangulation_options = triangulation_options
+        self._ba_optimizer = TwoViewBundleAdjustment(
+            reproj_error_thresholds=ba_reproj_error_thresholds,
+            robust_measurement_noise=True,
+            max_iterations=bundle_adjust_2view_maxiters,
+        )
+
+    def _augment_keypoints(self, keypoints_original: Keypoints, keypoints_new: Keypoints) -> Tuple[Keypoints, int]:
+        if len(keypoints_original) == 0:
+            return keypoints_new, 0
+        if len(keypoints_new) == 0:
+            return keypoints_original, 0
+
+        num_original_keypoints = len(keypoints_original)
+
+        # Note(Ayush): we avoid augmenting scales and responses and they do not make sense across different detectors
+        return (
+            Keypoints(coordinates=np.append(keypoints_original.coordinates, keypoints_new.coordinates, axis=0)),
+            num_original_keypoints,
+        )
+
+    def augment_correspondences(
+        self,
+        client: Client,
+        images: List[Future],
+        keypoints_list: List[Keypoints],
+        camera_intrinsics: List[gtsfm_types.CALIBRATION_TYPE],
+        two_view_outputs: Dict[Tuple[int, int], TWO_VIEW_OUTPUT],
+        gt_cameras: List[Optional[gtsfm_types.CAMERA_TYPE]],
+        gt_scene_mesh: Optional[Any],
+    ) -> Tuple[List[Keypoints], Dict[Tuple[int, int], TWO_VIEW_OUTPUT]]:
+        (
+            keypoints_to_augment_list,
+            putative_corr_idxs_to_augment,
+        ) = self._correspondence_generator.generate_correspondences(
+            client=client, images=images, image_pairs=list(two_view_outputs.keys())
+        )
+
+        def verify_putative_correspondences(
```

Comment:
Just curious, have you checked if we get similar accuracy from 

1. Merge unverified, and estimate 2-view w/ verification once.
2. Match and verify separately, then merge? 

My understanding was that https://www.kaggle.com/competitions/image-matching-challenge-2023/discussion/417407 did (1) not (2)
@travisdriver fyi

Reply:
Yes I saw that too. Let me work on approach 1 too then.

Reply:
Yes, I think (1) makes more sense